### ISSUE TITLE
Use Pulse API to store entries' fav state

### ISSUE THREAD

**Issue by mmmavis** [mozilla/network-pulse#310]
Currently Pulse uses `localStorage` to store a list of IDs of the entries that a user has fav'ed. Once the API is ready we should use the API to store this info on the database instead.

Question - Do we want to show heart(fav) icon ❤️ before users are logged in? If so, what does the icon do? Does clicking on it bring users to the login page? or?

/cc @kristinashu @Pomax @alanmoo

**Comment by alanmoo** [comment]
I think we discussed this, but can't quite remember where we ended up- was it that we'd store anonymous likes as a separate data point on the entries model?

**Comment by kristinashu** [comment]
Chatted with office peeps and we think it should bring users to the log in.

**Comment by Pomax** [comment]
there was a bit of a confusion about what hearts represented, too. Are they bookmarks (i.e. data that is strictly tied to users without an "aggregate" number and purely used to find previously bookmarked entries at some later point in time) or are they likes (data that links users and entries, and gives a signal to the poster and the community as a whole that an entry is meaningful to the community).

Depending on the social medium people are used to in their daily life, the heart icon will mean different things to different people, so we might need to examine exactly what role we want this toggle function to play, and whether we can possibly find more appropriate iconography.

(I'd personally lean towards treating "favoriting" an entry as "liking", with entries showing their likes-counts)

**Comment by kristinashu** [comment]
Just double checking, we are not counting how many times a post has been fav'ed, right? cc @xmatthewx

If we are counting, then lets stick with the heart but if we are not counting then lets open up a new ticket to mockup/discuss heart icon vs bookmark... either way, the icon shape is not a blocker.

**Comment by Pomax** [comment]
I don't know if we are right now in pulse, we've set up the model to at least do this (even if we don't end up exposing it, but if the idea is to also do a bit of community building, likes/upvotes are a great cheap way to achieve that)

**Comment by xmatthewx** [comment]
Let's stick with the icon and current behavior for now. We can follow up with some user testing later. 

The primary goal is to enable people to keep track of their favorites. Ideally, people can use the feature without login. Login enhances by syncing, but isn't required. 

If possible, let's keep Localstorage and check DB to sync latest info. We can resolve sync conflicts by keeping more favs. 

We do want to keep a count of favs to help us understand the network's interests. We don't currently have a plan to display the count. TBD.

**Comment by mmmavis** [comment]
I think I'm gonna work on this ticket next.

Technically wise I'm thinking

- tracks favourited(bookmarked) entries in localStorage: `localStorage.favedEntries` (array)
- every time user favourites an entry: 
    - we add the entry id to `localStorage.favedEntries`
    - we also initiate PUT Pulse API calls to`/entries/<id=number>/bookmark` for ids in `localStorage.favedEntries`. Once an entry's bookmarked status has been successfully updated in the db, we then remove the `id` from `localStorage.favedEntries`

@alanmoo @Pomax thoughts?

**Comment by Pomax** [comment]
That seems reasonable, at what point would you attempt the PUT operation?

**Comment by mmmavis** [comment]
> That seems reasonable, at what point would you attempt the PUT operation?

I was thinking every time after `localStorage.favedEntries` gets updated?

**Comment by Pomax** [comment]
conditional on the user being signed in though? Or just "regardless" and see if it goes through?

**Comment by mmmavis** [comment]
Right. I guess that also depends on if we want to prompt user to sign in or not. 

## For users who \*are not\* signed in

### Option 1

We store ids of their bookmarked entries in `localStorage` only. We don't prompt them to log in.

 ### Option 2

We store ids of their bookmarked entries in `localStorage` only as well as prompt them to log in. Once they are logged in, we make PUT calls for things in `localStorage.favedEntries`.

## For users who \*are\* signed in

Other than "after user clicks on the heart icon on an entry card on Pulse frontend", I'm not sure if we want to auto start the sync up process in the background. If so, when and how often?

**Comment by Pomax** [comment]
I can imagine option 2 to get really annoying after a while for people who inherently can't log in because they don't have a certified mozilla email account. I'd favour option 1, with a little notice only when an entry is faved and the localStorage list of faves is empty, to let them know that if they want these synced they'll need a log in and that in the mean time faves will be stored on a per-brower, per-device basis, so switching to a difference device or browser will not carry over any faves

**Comment by mmmavis** [comment]
Gonna wait for our decision in https://github.com/mozilla/network-pulse-api/issues/94 before I proceed.

**Comment by mmmavis** [comment]
@xmatthewx @kristinashu @alanmoo @Pomax 

As I'm working through this I started to wonder if it's good UX if we just sync `localStorage` bookmarks automatically in the background before asking/notifying user. There are also scenarios that I'm not sure how to handle. For example, say a user has entry `18` favourited in database, at some point she decided to log out, she favourited entry `18` again since that information was not stored in her browser (entry `18` appeared to be unfavourited after she logged out), and later she decided to unfavourite it... when she logs back in what do we wanna do with entry `18` in the database? Do we keep it favourited? Unfavourited?

Can I purpose that we do this in steps:

### As the first step, we don't do any sort of syncing

1. for users who are not signed it yet - we use `localStorage` to store their faved/bookmarked projects (these are remembered in the browser)

2. for users who are signed in - we use Pulse API to store their faved/bookmarked projects (these are remembered in the database)

3. We should have a clear explanation (probably somewhere on the /fav page) to explain how we are remembering users' bookmarks when they are not logged in vs logged in.

### later

And we can discuss how we wanna implement syncing and what the UX flow we wanna have. Bookmarks are currently stored by the time they were favourited (recently favourited first). We also need to decide how we wanna sort and display these bookmarked entries if we do wanna sync `localStorge` bookmarks with the database.

**Comment by xmatthewx** [comment]
Here's another option. Simpler than sync. What do you think?

When someone is logged in AND local storage favs are present, we...
 - don't display localstorage favs
 - display empty favs view help text
 - add a link to Import (see below)

This way, 
 - we only import, we don't sync. 
 - we only pull in current favs, we don't worry about unfavs. 

This seems simpler, good for users on multiple devices, and good for public or shared computers. And, it keeps favs useful for people who don't want or need to login. I'd suggest we clear localstorage favs on import to avoid confusion for people who logout. Some favs will be lost on shared computers, but that seems like a small risk, rare edge case. 

<img width="714" alt="screen shot 2017-05-22 at 5 47 46 pm" src="https://cloud.githubusercontent.com/assets/53411/26329787/cf2d953e-3f16-11e7-899a-d1b26f7c291a.png">

**Comment by Pomax** [comment]
I like this idea, @xmatthewx

**Comment by kristinashu** [comment]
Now that anyone can sign up, I think we should go back to plan A where tapping the heart prompts users to signin/up. This is how most apps and websites work. It avoids confusion for users around "I thought I already hearted this???". If someone doesn't want to sign up, then no big loss for them that they can't favorite thing. 

I don't want to block anything so move ahead however you see fit but want to to remind us of the original plan which I believe has the best UX.

**Comment by Pomax** [comment]
I kind of like an internet open enough not to need to log in just to fave something. Even if it's logging into a moz property. If there is no technical reason to, no logins win hands down, until you *need* to log in.

**Comment by xmatthewx** [comment]
> I think we should go back to plan A where tapping the heart prompts users to signin/up.

I thought about it. Wrote about it. Then deleted. Then thought about it again. 

We only have login for google users. And, most visitors will just be readers. They benefit little from login. If we can do this gracefully, it really fits our values.

**Comment by xmatthewx** [comment]
I talked with @hannahkane and @alanmoo about this again. It's a tough call. I think Kristina is totally right. I think Pomax is right too.  

I'd like to proceed with the import approach. `Reason 1`: love for the open web. `Reason 2`: we would need to design the whole tap heart, prompt login flow and all the branches. `Reason 3`: It's low risk for us to do something a bit experimental here. If this proves to be a mistake, now or later, we can switch gears. 

I will let @hannahkane and @kristinashu revise my copy and UI suggestion below (headed out on PTO). This might be better:

 > Did you add favs while logged out? [Import](#) them to your account. 

<img width="984" alt="screen shot 2017-05-23 at 2 00 47 pm" src="https://cloud.githubusercontent.com/assets/53411/26377393/0f73085a-3fde-11e7-90c1-7c1ba2d488fb.png">

**Comment by hannahkane** [comment]
I think @xmatthewx's revised copy works. No changes from me.

**Comment by kristinashu** [comment]
Yeah, works for me.

**Comment by mmmavis** [comment]
@kristinashu any suggestion on what we should show users after they click "import favs"? I was thinking displaying some sort of confirmation message without the use of modal dialog 😂

**Comment by kristinashu** [comment]
I guess we could replace the grey text that says "Did you add favs..." with something like "Your favs have been successfully imported."

**Comment by hannahkane** [comment]
@kristinashu - what do you think about simply "Your favs have been imported." I think "successfully" isn't needed.

**Comment by kristinashu** [comment]
Yes, much better!

**Comment by mmmavis** [comment]
@xmatthewx not sure if there's a Zenhub way to link this issue with the api issue but here is the api ticket https://github.com/mozilla/network-pulse-api/issues/118

**Comment by gideonthomas** [comment]
@mmmavis at the top of the issue below the description, you can "add dependency", maybe that's what you're looking for?
